karapace/utils.py

```diff
@@ -81,32 +81,42 @@ def ok(self):
         return 200 <= self.status_code < 300
 
 
+async def get_aiohttp_client():
+    import aiohttp
+    return aiohttp.ClientSession()
+
+
 class Client:
-    def __init__(self, server_uri=None, client=None):
+    def __init__(self, server_uri=None, client_factory=None):
         self.server_uri = server_uri or ""
         self.path_for = partial(urljoin, self.server_uri)
         self.session = requests.Session()
-        self.client = client
-        if self.client is None:
-            import aiohttp
-            self.client = aiohttp.ClientSession()
+        self.client_factory = get_aiohttp_client if client_factory is None else client_factory
```

Comment:
Why is the code no longer RAII?

Reply:
Client is created outside of an async context.
aiohttp >= 3.8.0 doesn't like that because it remembers the current async loop (or lack of loop) at init time and then complains loudly if the loop is different when the client is actually used. That's why I had to change it to lazy initialization.

Another option would be to reshape the main to have the `KafkaRest` class be initialized in an async context,
which means removing/reshaping `kc.run` (which actually calls `aiohttp.web.run_app`, which starts the async context and "should" be used directly as the main entry point of the program, according to [their doc](https://docs.aiohttp.org/en/stable/web_reference.html#aiohttp.web.run_app)):
https://github.com/aiven/karapace/blob/560a4ad64bfbf9ecbdcd7dd9be6f804bfc8a6986/karapace/kafka_rest_apis/__main__.py#L22-L24
Given all the inheritance happening around rapu, that might not be so easy, deferring only `_init_kafka_rest` could be an option:
https://github.com/aiven/karapace/blob/560a4ad64bfbf9ecbdcd7dd9be6f804bfc8a6986/karapace/kafka_rest_apis/__init__.py#L40-L46

Reply:
@kmichel-aiven thanks for the explanation. Maybe just add a comment in the code? I would only have figure this out when trying to make the code RAII and failing :sweat_smile: